relational_operators/BuildHashOperator.cpp

```diff
@@ -56,68 +54,174 @@ class TupleReferenceGenerator {
+                                                    QueryContext *query_context,
+                                                    StorageManager *storage_manager) {
+  DCHECK(placement_scheme_ != nullptr);
+  const std::size_t num_partitions =
+      input_relation_.getPartitionScheme().getPartitionSchemeHeader().getNumPartitions();
+  if (input_relation_is_stored_) {
+    for (std::size_t part_id = 0; part_id < num_partitions; ++part_id) {
+      for (const block_id input_block_id : input_relation_block_ids_in_partition_[part_id]) {
+        JoinHashTable *hash_table = query_context->getJoinHashTable(hash_table_group_index_, part_id);
         container->addNormalWorkOrder(
             new BuildHashWorkOrder(input_relation_,
                                    join_key_attributes_,
                                    any_join_key_attributes_nullable_,
                                    input_block_id,
                                    hash_table,
-                                   storage_manager),
+                                   storage_manager,
+                                   placement_scheme_->getNUMANodeForPartition(
+                                       input_relation_.getPartitionScheme().getPartitionForBlock(input_block_id))),
             op_index_);
       }
+    }
+  } else {
+    for (std::size_t part_id = 0; part_id < num_partitions; ++part_id) {
+      JoinHashTable *hash_table = query_context->getJoinHashTable(hash_table_group_index_, part_id);
+      while (num_workorders_generated_in_partition_[part_id] <
+             input_relation_block_ids_in_partition_[part_id].size()) {
+        block_id block_in_partition =
+            input_relation_block_ids_in_partition_[part_id][num_workorders_generated_in_partition_[part_id]];
+        container->addNormalWorkOrder(
+            new BuildHashWorkOrder(
+                input_relation_,
+                join_key_attributes_,
+                any_join_key_attributes_nullable_,
+                block_in_partition,
+                hash_table,
+                storage_manager,
+                placement_scheme_->getNUMANodeForPartition(
+                    input_relation_.getPartitionScheme().getPartitionForBlock(block_in_partition))),
+            op_index_);
+        ++num_workorders_generated_in_partition_[part_id];
+      }
+    }
+  }
+}
+#endif
+
+bool BuildHashOperator::getAllWorkOrders(WorkOrdersContainer *container,
+                                         QueryContext *query_context,
+                                         StorageManager *storage_manager,
+                                         const tmb::client_id scheduler_client_id,
+                                         tmb::MessageBus *bus) {
+  DCHECK(query_context != nullptr);
+
+  if (input_relation_is_stored_) {
+    if (!started_) {
+      if (input_relation_.hasPartitionScheme() && input_relation_.hasNUMAPlacementScheme()) {
```

Comment:
I think we should remove `hasNUMAPlacementScheme` from the condition. Imagine of the distributed case where we have partition info, but is not `NUMA` aware.

Also, I guess both `hasNUMAPlacementScheme` and `is_numa_aware_join_` are the same value, aren't they?